client/client.go

```diff
@@ -44,54 +45,52 @@ type Client struct {
 // It establishes a connection to the blockchain and verifies the integrity of contracts at the given address.
 // It uses the comm backend to initialize adapters for off-chain communication network.
 func NewEthereumPaymentClient(cfg Config, user perun.User, comm perun.CommBackend) (*Client, error) {
-	dialer, listener, err := initComm(comm, user.CommAddr)
-	if err != nil {
-		return nil, err
-	}
 	funder, adjudicator, err := connectToChain(cfg.Chain, user.OnChain)
 	if err != nil {
 		return nil, err
 	}
-	// Only off-chain account is unlocked. Accounts for the participant addresses
-	// will be unlocked by the client when required.
 	offChainAcc, err := user.OffChain.Wallet.Unlock(user.OffChain.Addr)
 	if err != nil {
 		return nil, errors.WithMessage(err, "off-chain account")
 	}
+	msgBus := net.NewBus(offChainAcc, comm.NewDialer())
 
-	c := client.New(offChainAcc, dialer, funder, adjudicator, user.OffChain.Wallet)
-	if err := loadPersister(c, cfg.DatabaseDir, cfg.PeerReconnTimeout); err != nil {
+	c, err := client.New(offChainAcc.Address(), msgBus, funder, adjudicator, user.OffChain.Wallet)
+	if err != nil {
+		return nil, errors.Wrap(err, "initializing state channel client")
+	}
+	if err = loadPersister(c, cfg.DatabaseDir, cfg.PeerReconnTimeout); err != nil {
 		return nil, err
 	}
+
 	client := &Client{
 		Client: c,
+		msgBus: msgBus,
 		wg:     &sync.WaitGroup{},
 	}
 
-	client.runAsGoRoutine(func() { client.Handle(&proposalHandler{}, &updateHandler{}) })
-	client.runAsGoRoutine(func() { client.Listen(listener) })
+	listener, err := comm.NewListener(user.CommAddr)
+	if err != nil {
+		return nil, err
+	}
+	client.runAsGoRoutine(func() { client.Handle(&ProposalHandler{}, &UpdateHandler{}) })
+	client.runAsGoRoutine(func() { msgBus.Listen(listener) })
 
 	return client, nil
 }
 
 // Close closes the client and waits until the listener and handler go routines return.
 //
-// Close depends on the following mechanisms implemented in client.Close:
-// 1. A context is passed to the infinite loop of handle function. When client.Close is called,
-//    this context is canceled, thereby causing the infinite loop in Handle function to stop.
-// 2. A callback to close the listener (listener.Close()) is registered to the client.Closer.
-//    When client.Close is called, this call back signals the listener to shutdown and hence
-//    stop the infinite loop in Listen function.
+// Close depends on the following mechanisms implemented in client.Close and bus.Close to signal the go-routines:
+// 1. When client.Close is invoked, it signals the infinite loop (via a context) in Handle function to stop.
+// 2. When bus.Close in invoked, it invokes EndpointRegistry.Close that shuts down the listener via onCloseCallback.
 func (c *Client) Close() error {
-	err := c.Client.Close()
+	err := errors.Wrap(c.Client.Close(), "closing channel client")
+	if busErr := c.msgBus.Close(); busErr != nil {
+		return errors.WithMessage(err, "closing message bus")
```

Comment:
`WithMessage` returns `nil` iff the passed error is `nil`. Hence a failure of closing the bus but not the client would return `nil`. Is this intended?

Reply:
No this was not intended. The method should return if  `c.Client.Close` returns error.

It is also good reason to add one more test case ;). Fixed in dbacd4225ec925f915e3727fd858e110c384dde1. See `client_test.go`. (Have moved the integration tests to `client_integ_test.go`.

I have added the commands to run this test in PR description.